pkg/cloudprovider/provider/aws/provider.go

```diff
@@ -578,10 +578,16 @@ func (p *provider) Create(machine *v1alpha1.Machine, data *cloudprovidertypes.Pr
 		return nil, awsErrorToTerminalError(err, "failed create instance at aws")
 	}
 
+	if pc.OperatingSystem == providerconfigtypes.OperatingSystemRHEL && config.manager != nil {
+		if err := rhsm.AddRHELSubscriptionFinalizer(machine, data.Update); err != nil {
```

Comment:
@moadqassem Why not using an admission controller to add the finalizer? I think it is cleaner and if my assumption that this is not provider specific holds it could avoid repetitions.

Reply:
Regarding this, I totally agree on adding this somewhere centralized where we can easily maintain it and avoid repetition, however, and what I saw as a stopper for this is: 

-  The RHSM is actually an optional manager so users can activate by providing an offline_token and if no offline_token is present then the manager will be off.
- Not all cloud providers support RHEL as an os take for e.g: Hetzner, DO and Packet. 
- If I start doing some filtration based on the cloud provider and act accordingly, I think this is a bad design because we are exposing a cloud provider-specific logic on a more generalized level which is the abstract machine level.

If those factors were not present I would have done it the way you suggested and having them centralized.

Reply:
See my comments below.

> Regarding this, I totally agree on adding this somewhere centralized where we can easily maintain it and avoid repetition, however, and what I saw as a stopper for this is:
> 
> * The RHSM is actually an optional manager so users can activate by providing an offline_token and if no offline_token is present then the manager will be off.
The fact of being optional does not impact its location.
> * Not all cloud providers support RHEL as an os take for e.g: Hetzner, DO and Packet.
This logic will apply to RHEL OS only, it's not a property of the cloud provider but of the OS. If a provider does not support this OS yet than the logic won't apply, I don't see problems here.
> * If I start doing some filtration based on the cloud provider and act accordingly, I think this is a bad design because we are exposing a cloud provider-specific logic on a more generalized level which is the abstract machine level.
Well, I think we both agree that conceptually this feature is OS specific and not provider specific. If we couple this logic to the provider because we think one day we could have a provider specific exception I don't think we're doing the best design ever.
> 
> If those factors were not present I would have done it the way you suggested and having them centralized.

Reply:
Well, I have been looking into it, over the weekend and done some changes regarding where those finalizers should be created and removed :).